### ISSUE TITLE
Expose unique constrains metadata to schema API and T4 model

### ISSUE THREAD

**Issue by filippobottega** [linq2db/linq2db#2037]
Hi,
  TT model generator sets correctly PrimaryKey attribute to properties of classes with attribute Table.
  Would it be possible to add also a new attribute, named `AlternateKey(string alternateKeyName)`, to identify properties that partecipate to a unique constraint?

  For example:

``` C#
	[Table(Schema="DWUSER", Name="ST01_COMPONENT")]
	public partial class ST01Component
	{
		[Column("ID_COMPONENT"),     PrimaryKey,  NotNull] public decimal IdComponent     { get; set; } // NUMBER
		[Column("SITE"),             AlternateKey("ST01_NC_CODE_SITE_NC_CODE_UN"),  Nullable] public string  Site            { get; set; } // NVARCHAR2(18)
		[Column("SFC"),              AlternateKey("ST01_NC_CODE_SITE_NC_CODE_UN"),  Nullable] public string  Sfc             { get; set; } // NVARCHAR2(384)
		[Column("ITEM"),                Nullable         ] public string  Item            { get; set; } // NVARCHAR2(384)
		[Column("ITEM_DESCRIPTION"),    Nullable         ] public string  ItemDescription { get; set; } // NVARCHAR2(120)
}
```

  The use case is the possibility to use this new attribute to generate new code based on it without the need to call the back end database another time.

  Thank you!

### Environment details
linq2db version: 3.0.0-preview.1
Database Server: Oracle
Database Provider: Oracle.ManagedDataAccess.Core (3.0.0-preview.1)
Operating system: Windows 10
.NET Framework:.NET Core 3.1

**Comment by MaceWindu** [comment]
To add new mapping attributes, we need a feature, that could benefit from it. Could you clarify how this attribute could be used by linq2db?

**Comment by filippobottega** [comment]
At the moment I think that linq2db doesn't need the new attribute.
I need it to automate code generation used to denormalize data at client side.
Another way could be to include in the generated code the database metadata in a way that were possible to the developer to consume in a new T4 model.
Letting the user to consume the DB metadata, read from linq2db, could be useful for the developer.

**Comment by MaceWindu** [comment]
In that case I think your best call is to add custom logic to T4 template to inspect and annotate loaded model with additional 3rd-party attribute.

If I understand it correctly, to do so, linq2db should provide unique constrains metadata? Like constrain name and used columns?

**Comment by filippobottega** [comment]
> In that case I think your best call is to add custom logic to T4 template to inspect and annotate loaded model with additional 3rd-party attribute.

Have you a guide or a quick suggestion to let me understand how to change linq2db T4 templates without the need to redo the changes when I upgrade or reinstall the linq2db's nuget package?

> If I understand it correctly, to do so, linq2db should provide unique constrains metadata? Like constrain name and used columns?

Exactly

**Comment by MaceWindu** [comment]
You need to add code to your `.tt` file (not to linq2db t4 templates) between model load and generation steps https://linq2db.github.io/articles/T4.html#example-of-generation-process-customization

and then go through Tables collection and add additional attributes to their columns.

**Comment by filippobottega** [comment]
Thank you for your suggested link, I have read the guide again and now understand that what linq2db let to do in the custom T4 templates is what I need.
The only missing point is the lack of unique index data in the section  **Useful members and data structures**.

**Comment by sdanyliv** [comment]
@filippobottega, unique constraints can give benefits only for query optimizations (removing unnecessary joins). Anyway for this case we have `HasUniqueKey` queryable extension.

**Comment by filippobottega** [comment]
If I well understand `HasUniqueKey` was created because the model lack of this information.
If you add unique keys to the model reading them from the DB, you could do query optimization without the need of `HasUniqueKey` queryable extension.

**Comment by sdanyliv** [comment]
You are right, a lack of time to implement. You can help us to retrieve Unique constraints by SchemaProvider. It can be good first PR.

Anyway `HasUniqueKey` helps to define unique constraint for the subquery, when you know which data returned by subquery.

**Comment by MaceWindu** [comment]
Consolidated from duplicate issue(s) — all three report the same root cause: linq2db doesn't load unique-constraint metadata into the database schema, so the scaffolder can't detect 1-1 relations and emits 1-many (`IEnumerable<T>`) associations.

### From #1339 (by @hardhub, 2018)

SQL Server: even with a unique index on the FK column, scaffolding generates `Relationship.OneToMany`. Maintainer note (@MaceWindu): "That's issue not about `Relation`, but about loading information about unique constrains, which allow to generate one-to-one form of association."

### From #4874 (by @advapiIT, 2025)

Concrete SQL Server repro — a `UNIQUE` constraint on the FK target column should make the association 1-1, but the scaffolder emits `IEnumerable<StaffAttribute>`. Maintainer confirmed (@MaceWindu): "This is a known issue. We currently don't load unique constraints into database schema required to detect 1-1 relations."

```sql
create table dbo.StaffAttributes
(
    Id            int identity constraint PK_StaffAttributes primary key,
    UserId        int not null
        constraint UQ_UserId unique
        constraint FK_User_StaffAttributes_UserId references dbo.[User] on delete cascade,
    DestinationId int not null
)
```

Scaffold CLI: `dotnet linq2db scaffold -p SqlServer --partial-entities true ...`
